test/src/unit-QueryCondition.cc

```diff
@@ -740,138 +1560,1121 @@ void test_apply(const Datatype type, bool var_size, bool nullable) {
+  result_tile.init_attr_tile(field_name);
+
+  test_apply_tile<T>(field_name, cells, type, array_schema, &result_tile);
+}
+
+TEST_CASE("QueryCondition: Test apply", "[QueryCondition][apply]") {
+  test_apply<int8_t>(Datatype::INT8);
+  test_apply<uint8_t>(Datatype::UINT8);
+  test_apply<int16_t>(Datatype::INT16);
+  test_apply<uint16_t>(Datatype::UINT16);
+  test_apply<int32_t>(Datatype::INT32);
+  test_apply<uint32_t>(Datatype::UINT32);
+  test_apply<int64_t>(Datatype::INT64);
+  test_apply<uint64_t>(Datatype::UINT64);
+  test_apply<float>(Datatype::FLOAT32);
+  test_apply<double>(Datatype::FLOAT64);
+  test_apply<char>(Datatype::CHAR);
+  test_apply<int64_t>(Datatype::DATETIME_YEAR);
+  test_apply<int64_t>(Datatype::DATETIME_MONTH);
+  test_apply<int64_t>(Datatype::DATETIME_WEEK);
+  test_apply<int64_t>(Datatype::DATETIME_DAY);
+  test_apply<int64_t>(Datatype::DATETIME_HR);
+  test_apply<int64_t>(Datatype::DATETIME_MIN);
+  test_apply<int64_t>(Datatype::DATETIME_SEC);
+  test_apply<int64_t>(Datatype::DATETIME_MS);
+  test_apply<int64_t>(Datatype::DATETIME_US);
+  test_apply<int64_t>(Datatype::DATETIME_NS);
+  test_apply<int64_t>(Datatype::DATETIME_PS);
+  test_apply<int64_t>(Datatype::DATETIME_FS);
+  test_apply<int64_t>(Datatype::DATETIME_AS);
+  test_apply<char*>(Datatype::STRING_ASCII);
+  test_apply<char*>(Datatype::STRING_ASCII, true);
+  test_apply<char*>(Datatype::STRING_ASCII, false, true);
+}
+
+TEST_CASE(
+    "QueryCondition: Test combinations", "[QueryCondition][combinations]") {
+  const std::string field_name = "foo";
+  const uint64_t cells = 10;
+  const Datatype type = Datatype::UINT64;
+
+  // Initialize the array schema.
+  ArraySchema array_schema;
+  Attribute attr(field_name, type);
+  REQUIRE(
+      array_schema.add_attribute(make_shared<Attribute>(HERE(), &attr)).ok());
+  Domain domain;
+  Dimension dim("dim1", Datatype::UINT32);
+  uint32_t bounds[2] = {1, cells};
+  Range range(bounds, 2 * sizeof(uint32_t));
+  REQUIRE(dim.set_domain(range).ok());
+  REQUIRE(domain.add_dimension(make_shared<Dimension>(HERE(), &dim)).ok());
+  REQUIRE(array_schema.set_domain(make_shared<Domain>(HERE(), &domain)).ok());
+
+  // Initialize the result tile.
+  ResultTile result_tile(0, 0, array_schema);
+  result_tile.init_attr_tile(field_name);
+  ResultTile::TileTuple* const tile_tuple = result_tile.tile_tuple(field_name);
+  Tile* const tile = &std::get<0>(*tile_tuple);
+
+  // Initialize and populate the data tile.
+  REQUIRE(tile->init_unfiltered(
+                  constants::format_version,
+                  type,
+                  cells * sizeof(uint64_t),
+                  sizeof(uint64_t),
+                  0)
+              .ok());
+  uint64_t* values = static_cast<uint64_t*>(malloc(sizeof(uint64_t) * cells));
+  for (uint64_t i = 0; i < cells; ++i) {
+    values[i] = i;
+  }
+  REQUIRE(tile->write(values, 0, cells * sizeof(uint64_t)).ok());
+
+  SECTION("Basic AND condition.") {
```

Comment:
Do you want the tests to be split per query condition or to for there to be a test per query condition and per function? So for one query condition, there are 3 separate test cases, one for apply, apply_sparse, apply_dense. Personally, I think it's fine to have one test case per query condition, since the setup is the same (so this would be similar to the integration tests). What do you think?